File: benchmark/quicksort.go
package main

import "fmt"

func partition(arr []int64, low, high int64) int64 {
	pivot := arr[high]
	i := low - 1
	for j := low; j < high; j++ {
		if arr[j] < pivot {
			i++
			arr[i], arr[j] = arr[j], arr[i]
		}
	}
	arr[i+1], arr[high] = arr[high], arr[i+1]
	return i + 1
}

func quicksort(arr []int64, low, high int64) {
	if low < high {
		p := partition(arr, low, high)
		quicksort(arr, low, p-1)
		quicksort(arr, p+1, high)
	}
}

func runSort(size int64) int64 {
	arr := make([]int64, size)
	var next int64 = 42
	for i := range arr {
		next = next*1664525 + 1013904223
		arr[i] = next % 100000
	}
	quicksort(arr, 0, size-1)
	return arr[size/2]
}

func main() {
	fmt.Printf("Median checksum: %d\n", runSort(1000000))
}

File: benchmark/quicksort.rs
fn partition(arr: &mut [i64], low: i64, high: i64) -> i64 {
    let pivot = arr[high as usize];
    let mut i = low - 1;
    for j in low..high {
        if arr[j as usize] < pivot {
            i += 1;
            arr.swap(i as usize, j as usize);
        }
    }
    arr.swap((i + 1) as usize, high as usize);
    i + 1
}

fn quicksort(arr: &mut [i64], low: i64, high: i64) {
    if low < high {
        let p = partition(arr, low, high);
        quicksort(arr, low, p - 1);
        quicksort(arr, p + 1, high);
    }
}

fn run_sort(size: i64) -> i64 {
    let mut arr = vec![0i64; size as usize];
    let mut next: i64 = 42;
    for slot in &mut arr {
        next = next.wrapping_mul(1664525).wrapping_add(1013904223);
        *slot = next % 100000;
    }
    quicksort(&mut arr, 0, size - 1);
    arr[(size / 2) as usize]
}

fn main() {
    println!("Median checksum: {}", run_sort(1000000));
}
